swap/src/protocol/alice/event_loop.rs

```diff
@@ -153,14 +156,14 @@ where
                     match swarm_event {
                         SwarmEvent::Behaviour(OutEvent::ExecutionSetupStart { peer, btc, xmr }) => {
 
-                            let tx_redeem_fee = self.bitcoin_wallet
+                            let tx_redeem_fee = self.internal_bitcoin_wallet
                                 .estimate_fee(bitcoin::TxRedeem::weight(), btc)
                                 .await;
-                            let tx_punish_fee = self.bitcoin_wallet
+                            let tx_punish_fee = self.internal_bitcoin_wallet
                                 .estimate_fee(bitcoin::TxPunish::weight(), btc)
                                 .await;
-                            let redeem_address = self.bitcoin_wallet.new_address().await;
-                            let punish_address = self.bitcoin_wallet.new_address().await;
+                            let redeem_address = self.personal_bitcoin_wallet.new_address().await;
+                            let punish_address = self.personal_bitcoin_wallet.new_address().await;
```

Comment:
Passing around two instead of one wallet is quite the invasive change. What if we instead introduce functions such as `new_redeem_address` and `new_punish_address` and decide internally within `bitcoin::Wallet`, where we derive these addresses from? I think that would be less invasive because it allows us to handle things within `bitcoin::Wallet`.

Reply:
I thought about something similar but decided against it. 
Here are my thoughts: 

1. I thought about adding a new field to `bitcoin::Wallet` i.e. `personal_wallet` so that we can achieve what you proposed. I decided against it because it would destroy the meaning of `bitcoin::Wallet`, because it's not a single `Wallet` anymore but something like `Wallets`... the abstraction did not add-up for me because how do you know from the outside what do you call `balance()`/`sign`/`new_address` on without knowing the internal structure in advance? 
2. I then wanted to have a completely separate wallet implementation for the read-only wallet, i.e. independent from `bitcoin::wallet`. For example, we could have `{crate}::bitcoin::wallets::internal` and `{crate}::bitcoin::wallets::personal`. `personal` could than be an offline-only wallet allowing easier testing key and address derivations. I then ran into problems because I had to assert the balances in the tests. It would also add some duplicated code.
3. I decided to re-use the same wallet which I think turned out quite nice. They are too similar to separate and it allowed me to not have to implement a new wallet for e2e tests which would allow me to assert on the balances. 

Imho this way is best trade-off between asking developers to go yet one level further down.

Reply:
> I thought about adding a new field to bitcoin::Wallet i.e. personal_wallet so that we can achieve what you proposed. I decided against it because it would destroy the meaning of bitcoin::Wallet, because it's not a single Wallet anymore but something like Wallets... the abstraction did not add-up for me because how do you know from the outside what do you call balance()/sign/new_address on without knowing the internal structure in advance?

Do you have to know that? I don't see why you would.

Unless I am missing something, we would only need to remove `new_address` and add new APIs that qualify, what the semantic of this address is:

- `new_redeem_address`
- `new_refund_address`

Like I pointed out in the overall feedback comment above, our `bitcoin::Wallet` is not a general-purpose Bitcoin wallet (unlike `bdk::Wallet`). It is more the wallet component for a swap application. Anything more generic can still live in there obviously but doesn't need to hold us back in creating an abstraction layer that works _better_ for our swap protocol with better being defined as: the protocol doesn't need to care about application semantics like the xpub feature.

We don't have to treat `bitcoin::Wallet` as a single wallet, that is what an instance of `bdk::Wallet` is for. Instead, imagine `bitcoin::Wallet` being a trait and serving as the interface for our protocol to access wallet functionality. What happens under the hood is irrelevant to the protocol as long as the functionality offered is internally consistent, f.e. `balance` doesn't return coins that you can't actually spend.

In the current design, we introduce a lot of complexity by "leaking" into the protocol that some clients (only ASB for now) might want a different wallet for redeem/refund/punish.